pyemu/utils/helpers.py

```diff
@@ -235,6 +235,137 @@ def autocorrelated_draw(pst,struct_dict,time_distance_col="distance",num_reals=1
+                    # select pars to use specsim for
+                    gr_df = par_df.loc[grd_p]
+                    gr_df = gr_df.astype({"i": int, "j": int})  # make sure int
+                    # (won't be if there were nans in concatenated df)
+                    if len(gr_df) > 0:
+                        # get specsim object for this geostruct
+                        ss = pyemu.geostats.SpecSim2d(
+                            delx=delr,
+                            dely=delc,
+                            geostruct=geostruct,
+                        )
+                        # specsim draw (returns df)
+                        gr_pe1 = ss.grid_par_ensemble_helper(
+                            pst=pst,
+                            gr_df=gr_df,
+                            num_reals=num_reals,
+                            sigma_range=sigma_range,
+                            logger=logger,
+                        )
+                        # append to list of specsim drawn pars
+                        gr_pe_l.append(gr_pe1)
+                        # rebuild struct_dict entry for this geostruct
+                        # to not include specsim pars
+                        struct_dict[geostruct] = []
+                        # loop over all in list associated with geostruct
+                        for p_df in par_df_l:
+                            # if pars are not in the specsim pars just created
+                            # assign them to this struct_dict entry
+                            # needed if none specsim pars are linked to same geostruct
+                            if not p_df.index.isin(gr_df.index).all():
+                                struct_dict[geostruct].append(p_df)
+                            else:
+                                subset = subset.difference(p_df.index)
+        if len(gr_pe_l) > 0:
+            gr_par_pe = pd.concat(gr_pe_l, axis=1)
+        logger.log("spectral simulation for grid-scale pars")
+    # draw remaining pars based on their geostruct
+    if not subset.empty:
+        logger.log(f"Drawing {len(subset)} non-specsim pars")
+        pe = pyemu.helpers.geostatistical_draws(
+            pst,
+            struct_dict=struct_dict,
+            num_reals=num_reals,
+            sigma_range=sigma_range,
+            scale_offset=scale_offset,
+            subset=subset
+        )
+        logger.log(f"Drawing {len(subset)} non-specsim pars")
+        if gr_par_pe is not None:
+            logger.log(f"Joining specsim and non-specsim pars")
+            exist = gr_par_pe.columns.intersection(pe.columns)
+            pe = pe._df.drop(exist, axis=1)  # specsim par take precedence
+            pe = pd.concat([pe, gr_par_pe], axis=1)
+            pe = pyemu.ParameterEnsemble(pst=pst, df=pe)
+            logger.log(f"Joining specsim and non-specsim pars")
+    else:
+        pe = pyemu.ParameterEnsemble(pst=pst, df=gr_par_pe)
+    logger.log("drawing realizations")
+    return pe.copy()
```

Comment:
note to self: work out why we were returning a copy here (probs a pandas ref vs copy thing).